linger-users: Fix strict shell checks, fix for mutable users

me-and opened the pull request:
NixOS 24.11 introduced `systemd.enableStrictShellChecks`, which, when enabled, complains about parsing `ls` output in the systemd unit to manage lingering users. This PR rewrites the script involved to avoid those warnings.

While fixing this, I also noticed and fixed two bugs:

- One part of the tests for systemd user lingering would always pass, because it was setting up a user with UID 10001 and then checking there was no user slice for a user with UID 1001.
- If a system had systemd config that marked users as lingering, but the NixOS configuration had no users configured to linger (perhaps because some users were previously configured to linger but that configuration was removed), the systemd configuration wouldn't get cleaned up, so those users would continue to linger.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

me-and committed:
linger-users: fix shellcheck warnings

Running with systemd.enableStrictShellChecks with lingering users causes
failures due to parsing the output from `ls`.  Rewrite the script to
avoid parsing ls, and instead rely on loginctl enable-linger and
disable-linger commands being idempotent and run them unconditionally.

This also fixes a bug where the systemd unit for adding and removing
lingering user configuration is only enabled if there are users
configured with lingering in the NixOS configuration.  This means that
if a NixOS system is built with some lingering users, then the linger
configuration is removed from all those users, the script to disable
lingering won't be run, and those users will incorrectly continue to
have lingering enabled.

Fixes #418101.

me-and commented:
(Created as a draft while I do some more testing)

me-and committed:
tests/systemd-user-linger: correct uid

Make the UIDs for test user "bob" consistent to avoid potential false
positive test results.

This test previously set up a user with UID 10001 then checked to make
sure no user slice for a user with UID 1001 existed.  That would always
be true, since there is no user with UID 1001.

me-and committed:
tests/systemd-user-linger: test clearing lingering users

Split the existing system-user-linger tests into two: one which tests
lingering is set up for the correct users and only the correct users,
and one which checks that lingering is cleared for users that don't
exist.

This means that the latter test now also ensures the lingering
configuration is handled even if there are no users configured who
should be lingering, where previously such lingering configuration
wouldn't be touched.

me-and commented:
@ofborg test systemd-user-linger

me-and commented:
Requested reviews from @jwillikers and @adamcstephens as people who seem to have played with this code and function recently.

me-and commented:
@ofborg test systemd-user-linger

me-and commented:
Problem with this approach I've just spotted: it assumes `users.mutableUsers = false`. It should be possible for someone who has set `users.mutableUsers = true` (or left it to default to `true`) to add user accounts and set them to linger without them being overwritten by `linger-users.service`.

me-and committed:
tests/systemd-user-linger-purge: check mutable user handling

Mutable users configured to linger shouldn't have their lingering
deconfigured by the systemd unit for managing lingering users.

Test in both the scenario where there are no users configured in NixOS
to linger, and where there are such users, to catch the case at
https://github.com/NixOS/nixpkgs/pull/363209#issuecomment-2987448905
where mutable lingering users would be incorrectly cleared if there were
any immutable lingering users.

me-and committed:
users-groups: use loginctl from configured systemctl

If a user has configured a different systemd package,
linger-users.service should respect that and use the provided loginctl
executable rather than the one from the default nixpkgs package.

me-and commented:
That now fixed: users configured manually outside NixOS configuration won't have their lingering settings changed by the NixOS linger-users.service unit.

I've also spotted and fixed a minor bug where a user's `systemd.package` configuration isn't respected by linger-users.service.

me-and commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 363209 --no-shell --print-result -p nixosTests.systemd-user-linger -p nixosTests.systemd-user-linger-purge`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 tests built:</summary>
  <ul>
    <li>nixosTests.systemd-user-linger</li>
    <li>nixosTests.systemd-user-linger-purge</li>
  </ul>
</details>

LordGrimmauld commented on nixos/modules/config/users-groups.nix:
```diff
@@ -893,29 +893,46 @@ in
+                # Use `id` to check if the user exists rather than checking the
+                # NixOS configuration, as it may be that the user has been
+                # manually configured, which is permitted if users.mutableUsers
+                # is false (the default).
```
```suggestion
                # is true (the default).
```

This one is just inverted.

LordGrimmauld commented on nixos/modules/config/users-groups.nix:
```diff
@@ -893,29 +893,46 @@ in
+            fi
+            if (( ''${#lingeringUserNames[*]} > 0 )); then
+                ${pkgs.systemd}/bin/loginctl enable-linger "''${lingeringUserNames[@]}"
+            fi
```
```suggestion
            if (( ''${#nonLingeringUserNames[*]} > 0 )); then
                ${lib.getExe' config.systemd.package "loginctl"} "''${nonLingeringUserNames[@]}"
            fi
            if (( ''${#lingeringUserNames[*]} > 0 )); then
                ${lib.getExe' config.systemd.package "loginctl"} enable-linger "''${lingeringUserNames[@]}"
            fi
```

Hardcoding `pkgs.systemd` here means `systemd.package` overrides would be ignored. We should use the systemd package from config instead.

LordGrimmauld replied:
That commit is not in here, and your current changes would very much conflict if that is in some other PR. But yes, if you want to pick that commit into here, go ahead.

LordGrimmauld reviewed (changes requested):
Not super familiar with the whole linger stack, but i do still have some thoughts on this

LordGrimmauld commented:
Oh and you probably need to target this PR against the new-ish `staging-nixos` branch to bundle it together with kernel update. Makes hydra load more manageable.

me-and committed:
linger-users: use systemd directory options

Using systemd properties avoids the need for manually running mkdir and
cd commands, and helps systemd clean up properly when appropriate.

Suggested-By: Grimmauld <grimmauld@grimmauld.de>

me-and committed:
linger-users: log when deconfiguring linger

If systemd has recorded that a user should be lingering despite them not
having an account on the system, that record is removed.  When that
happens, log for the sake of future debugging and investigations.

Suggested-By: Grimmauld <grimmauld@grimmauld.de>

me-and commented:
> Oh and you probably need to target this PR against the new-ish `staging-nixos` branch to bundle it together with kernel update. Makes hydra load more manageable.

Ah, yes. I don't think that branch existed when I created this PR, but I'll re-target now.

LordGrimmauld reviewed (changes requested):
Do you mind rebasing onto `git merge-base origin/master origin/nixos-staging` and retargetting the PR to `nixos-staging`? This will rebuild all nixos tests, becuase the module became unconditional.

LordGrimmauld reviewed (commented):
rebase onto a bit less crusty nixpkgs (and fixing fmt) would be nice, but i don't see any immediate obvious issues just from reading code anymore. Thank you!

LordGrimmauld commented:
the nixpkgs used here is from 2024, testing against *that* is less than ideal.

LordGrimmauld reviewed (approved):
I don't see any issues anymore, but i want a second set of eyes (preferably systemd team) before merging. Thank you for tackling this!

LordGrimmauld commented:
One more thought (and cc @nikstur for bashless): this now adds a script-based service back into the default system closure. Even worse, there is no good way to disable this service, breaking features such as the [bashless profile](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/profiles/bashless.nix). We should probably expose a way to force-disable linger (and set that option in the bashless profile)

me-and committed:
linger-users: allow disabling for bashless profile

The linger-users systemd unit runs a Bash script.  To allow this to be
avoided for the bashless profile, provide an option to have NixOS not
manage lingering for any users.

To make this feasible, add the possibility for each individual user
account to not have its lingering configuration managed by NixOS at all,
and make this the default from 26.05.  In practice, this won't result in
a change of behaviour except for people who manually use `loginctl
enable-linger` commands to add lingering for some user accounts, then
rely on NixOS to disable lingering the next time the systemd units are
restarted.

me-and commented:
@LordGrimmauld @nikstur huh, I hadn't realised there was a bashless profile!

I've just added another commit that adds a `users.manageLingering` toggle, which I believe addresses this concern. I'm a bit wary about scope creep, but I can't think of a better way to do this without breaking the bashless profile.

LordGrimmauld reviewed (approved):
Thank you! I think this is mergable before 25.11, but i want a second set of eyes by either some systemd people or nikstur

nikstur reviewed (commented):
Was just about to review :upside_down_face: too slow

LordGrimmauld commented:
Oh damn, sorry! I can do these in a follow-up if you want.
